Source/Core/Common/Keyboard.cpp

```diff
@@ -0,0 +1,326 @@
+// Copyright 2025 Dolphin Emulator Project
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include "Common/Keyboard.h"
+
+#include <array>
+#include <map>
+#include <mutex>
+#include <utility>
+
+#ifdef HAVE_SDL2
+#include <SDL_events.h>
+#include <SDL_keyboard.h>
+
+// Will be overridden by Dolphin's SDL InputBackend
+u32 Common::KeyboardContext::s_sdl_init_event_type(-1);
+u32 Common::KeyboardContext::s_sdl_update_event_type(-1);
+u32 Common::KeyboardContext::s_sdl_quit_event_type(-1);
+#endif
+
+#include "Core/Config/MainSettings.h"
+#include "Core/Config/SYSCONFSettings.h"
+#include "DiscIO/Enums.h"
+
+namespace
+{
+// Translate HID usage ID based on the host and the game keyboard layout:
+//  - we need to take into account the host layout as we receive raw scan codes
+//  - we need to consider the game layout as it might be different from the host one
+u8 TranslateUsageID(u8 usage_id, int host_layout, int game_layout)
+{
+  if (host_layout == game_layout)
+    return usage_id;
+
+  // Currently, the translation is partial (i.e. alpha only)
+  if (usage_id != Common::HIDUsageID::M_AZERTY &&
+      (usage_id < Common::HIDUsageID::A || usage_id > Common::HIDUsageID::Z))
+  {
+    return usage_id;
+  }
+
+  switch (host_layout | game_layout)
+  {
+  case Common::KeyboardLayout::AZERTY_QWERTZ:
+  {
+    static const std::map<u8, u8> TO_QWERTZ{
+        {Common::HIDUsageID::A_AZERTY, Common::HIDUsageID::A},
+        {Common::HIDUsageID::Z_AZERTY, Common::HIDUsageID::Z_QWERTZ},
+        {Common::HIDUsageID::Y, Common::HIDUsageID::Y_QWERTZ},
+        {Common::HIDUsageID::Q_AZERTY, Common::HIDUsageID::Q},
+        {Common::HIDUsageID::M_AZERTY, Common::HIDUsageID::M},
+        {Common::HIDUsageID::W_AZERTY, Common::HIDUsageID::W},
+        {Common::HIDUsageID::M, Common::HIDUsageID::M_AZERTY},
+    };
+    static const std::map<u8, u8> TO_AZERTY{
+        {Common::HIDUsageID::Q, Common::HIDUsageID::Q_AZERTY},
+        {Common::HIDUsageID::W, Common::HIDUsageID::W_AZERTY},
+        {Common::HIDUsageID::Z_QWERTZ, Common::HIDUsageID::Z_AZERTY},
+        {Common::HIDUsageID::A, Common::HIDUsageID::A_AZERTY},
+        {Common::HIDUsageID::M_AZERTY, Common::HIDUsageID::M},
+        {Common::HIDUsageID::Y_QWERTZ, Common::HIDUsageID::Y},
+        {Common::HIDUsageID::M, Common::HIDUsageID::M_AZERTY},
```

Comment:
Not that this is a hot path, but I wonder if just using arrays like the `QWERTY_AZERTY` translation would be more performant than `std::map`.

Reply:
I'm not sure what would be the optimal way to test it but if I convert this `map`+`find` logic to an `array<pair>`+`range-for` (similar to `BI_MAP` below), it seems to degrade the emulation speed? It went (with unlimited CPU speed) from 390\~400% to 380\~390% when typing the characters needing translation.

I don't think that's the ideal way to benchmark it as it probably depends on my PC spec and might raise different results with another PC or another compiler. I went for the `std::map` because I assumed its `find`'s complexity was probably better than a range-for loop. Then, for the bi_map case, we didn't have an appropriate container+algorithm readily available that'd fit my use case (afaict) which is why I used an array instead.

Reply:
I wouldn't expect us to see a performance difference either way here. It's fine as-is.